validator/src/validator.rs

```diff
@@ -232,6 +237,50 @@ where
         self.init_block_producer(head_hash);
     }
 
+    fn push_block_timestamp(&mut self, timestamp: u64) {
+        // Only push the timestamps once caught up with the chain
+        if !self.is_synced() {
+            return;
+        }
+
+        let amount_of_timestamps = self.recent_block_timestamps.len();
+        if amount_of_timestamps >= 30 {
+            // The VeqDeque should not grow unbounded
+            assert!(amount_of_timestamps < 31);
+
+            self.recent_block_timestamps.pop_front();
+        }
+
+        self.recent_block_timestamps.push_back(timestamp);
+    }
+
+    fn compute_producer_timeout(&self) -> u64 {
+        let amount_of_timestamps = self.recent_block_timestamps.len();
+
+        // If 0 or 1 timestamp have been stored, simply return the Policy::BLOCK_PRODUCER_TIMEOUT
+        if amount_of_timestamps <= 1 {
+            return Policy::BLOCK_PRODUCER_TIMEOUT;
+        }
+
+        let mut summed_durations = 0;
+        for i in 1..amount_of_timestamps {
+            // Compute the difference between the timestamp (i) and it's immediate predecessor (i - 1)
+            summed_durations +=
+                self.recent_block_timestamps[i] - self.recent_block_timestamps[i - 1];
+        }
+
+        let average_duration: u64 = summed_durations / ((amount_of_timestamps as u64) - 1);
```

Comment:
```suggestion
        let complete_duration = self.recent_block_timestamps[amount_of_timestamps - 1] - self.recent_block_timestamps[0];
        let average_duration: u64 = complete_duration / ((amount_of_timestamps as u64) - 1);
```

Reply:
This is only true if the blocks are in order. The presence of skip blocks makes this a non guaranteed property.

Reply:
> This is only true if the blocks are in order.

This isn't true. The change I proposed doesn't need to assign any semantic value to the elements in the list.

You're computing `(a[1] - a[0]) + (a[2] - a[1]) + … + (a[n] - a[n-1])`, reordering, we get `(a[n] - a[n-1]) + (a[n-1] - a[n-2]) + … + (a[2] - a[1]) + (a[1] - a[0])`, associativity `a[n] + (-a[n-1] + a[n-1]) + … + (-a[1] + a[1]) - a[0]`, so `a[n] + 0 + 0 + … + 0 + a[0] = a[n] - a[0]`.

So if I got everything correctly, and if you think my formula is wrong, then the current formula is wrong, too.

Reply:
Apologies, yes i meant the marked section to already be wrong, as blocks are not necessarily pushed in ascending timestamp order. 

After a little more thought it should not be a problem though as long as rebranches get worked into this, as that should restore that order. Depending on what guarantees are given for what blocks are at least in there that is not a problem, but if there are no guarantees `summed_duration` might even get negative because of it. 

Importantly I don't see that the first block is necessarily on the same fork as the last block.

Your change obviously has nothing t do with that, I guess I should have created a new comment here instead of adding to yours.